test/sharness/t0110-gateway.sh

```diff
@@ -22,60 +22,197 @@ apiport=$PORT_API
 # define a function to test a gateway, and do so for each port.
 # for now we check 5001 here as 5002 will be checked in gateway-writable.
 
-test_expect_success "GET IPFS path succeeds" '
-  echo "Hello Worlds!" >expected &&
-  HASH=$(ipfs add -q expected) &&
-  curl -sfo actual "http://127.0.0.1:$port/ipfs/$HASH"
-'
+TEST_COMMAND=test_expect_success
 
-test_expect_success "GET IPFS path output looks good" '
-  test_cmp expected actual &&
-  rm actual
-'
+# Overwrites: actual.headers, actual.
+test_run_curl() {
+  rm -f actual.headers actual
+  curl -s -D actual.headers -o actual "$@"
+}
 
-test_expect_success "GET IPFS path on API forbidden" '
-  test_curl_resp_http_code "http://127.0.0.1:$apiport/ipfs/$HASH" "HTTP/1.1 403 Forbidden"
-'
+test_actual_headers_http_status() {
+  test_should_contain '^HTTP/1.1 '"$1"'.$' actual.headers
+}
+
+test_get_max_age() {
+  perl -nwe 's/^Cache-Control:.* max-age=([0-9]+).*$/$1/ && print' "$1"
+}
+
+# Overwrites: actual.max-age, expected.max-age.
+test_expect_max_age() {
+  test_get_max_age actual.headers >actual.max-age
+  printf "%s\n" "$1" >expected.max-age
+  test_cmp expected.max-age actual.max-age || test_fsh cat actual.headers
+}
+
+# Overwrites: actual.max-age, expected.max-age.
+test_expect_no_max_age() {
+  test_get_max_age actual.headers >actual.max-age
+  : >expected.max-age
+  test_cmp expected.max-age actual.max-age || test_fsh cat actual.headers
+}
+
+test_item() {
+  _item="$1"
+  _url="$2"
+  _http_status="$3"
+  _expected_output_file="$4"
+  _max_age="$5"
+  _etag="$6"  # Without the surrounding quote marks
+
+  "$TEST_COMMAND" "$_item: responds with $_http_status" '
+    test_run_curl "$_url" &&
+    test_actual_headers_http_status "$_http_status"
+  '
+
+  case "$_expected_output_file" in
+    # Make sure a skipped output test is conspicuous at the call site.
+    IGNORE_OUTPUT) ;;
+    *)
+      "$TEST_COMMAND" "$_item: output looks good" '
+        test_cmp "$_expected_output_file" actual
+      '
+      ;;
+  esac
+
+  case "$_max_age" in
+    IGNORE_MAX_AGE) ;;
+    NO_MAX_AGE)
+      "$TEST_COMMAND" "$_item: has no Cache-Control max-age" '
+        test_expect_no_max_age
+      '
+      ;;
+    *)
+      "$TEST_COMMAND" "$_item: has Cache-Control max-age=$_max_age" '
+        test_expect_max_age "$_max_age"
+      '
+      ;;
+  esac
+
+  case "$_etag" in
+    NO_ETAG)
+      "$TEST_COMMAND" "$_item: has no ETag" '
+        test_should_not_contain "^[Ee][Tt][Aa][Gg]: " actual.headers
+      '
+      ;;
+    *)
+      # The go http package replaces ETag with Etag.  Accept both.  (HTTP
+      # header names are case-insensitive but ETag is the official name.)
+      "$TEST_COMMAND" "$_item: has ETag" '
```

Comment:
dont do this, this leads to macro hell fast. just test both, write both out. it's much clearer and waaay easier to instrument + debug.